core/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -223,19 +224,21 @@ private void ensureNoPre019State() throws Exception {
      * index metadata with new settings.
      */
     private void pre20Upgrade() throws Exception {
-        MetaData metaData = loadMetaState();
-        List<IndexMetaData> updateIndexMetaData = new ArrayList<>();
-        for (IndexMetaData indexMetaData : metaData) {
+        List<PersistedIndexMetaData> metaData = metaStateService.loadIndicesStatesNoFolderNameValidation();
```

Comment:
We can use `GatewayMetaState#loadMetaState` here, it uses `MetaStateService#loadFullState`, which loads the global state along with all the `IndexMetaData`. We can just iterate on the loaded index meta data and upgrade? we can use the `clusterUUID` from the global state here, since we are just upgrading or make the `MetaData.Builder` aware of `PersistedIndexMetaData` (don't know if this is worth it?)